[Docs] Add an ionogram example in MMYOLO application (#643)
open-mmlab/mmyolo: docs/en/recommended_topics/application_examples/ionogram_detection.md

<div align=center>
<img width="40%" src="https://user-images.githubusercontent.com/67947949/223638535-c4583d88-aa5a-4f21-b35a-e6e8328c9bd4.jpg"/>

Preview of annotated images

</div>

1. Dataset prepration

After downloading the data, put it in the root directory of the MMYOLO repository, and use `unzip test.zip` (for Linux) to unzip it to the current folder. The structure of the unzipped folder is as follows:

```shell
Iono4311/
├── images
|      ├── 20130401005200.png
|      └── ...
└── labels
       ├── 20130401005200.json
       └── ...
```

The `images` directory contains input images，while the `labels` directory contains annotation files generated by labelme.

2. Convert the dataset into COCO format

Use the script `tools/dataset_converters/labelme2coco.py` to convert labelme labels to COCO labels.

```shell
python tools/dataset_converters/labelme2coco.py --img-dir ./Iono4311/images \
                                                --labels-dir ./Iono4311/labels \
                                                --out ./Iono4311/annotations/annotations_all.json
```

> 3. Check the converted COCO label

Reviewer: ```suggestion
3. Check the converted COCO labels
```

> Use the following command to display the COCO labels on the image, which verifies that the conversion has been problem-free.

Reviewer: ```suggestion
To confirm that the conversion process went successfully, use the following command to display the COCO labels on the images.
```

├── annotations
│   ├── annotations_all.json
│   ├── class_with_id.txt
│   ├── test.json
│   ├── train.json
│   └── val.json
├── classes_with_id.txt
├── images
├── labels
├── test_images
├── train_images
└── val_images
```

## Config files

The configuration files are stored in the directory `/projects/misc/ionogram_detection/`.

1. Dataset analysis

To perform a dataset analysis, a sample of 200 images from the dataset can be analyzed using the `tools/analysis_tools/dataset_analysis.py` script.

```shell
python tools/analysis_tools/dataset_analysis.py projects/misc/ionogram_detection/yolov5/yolov5_s-v61_fast_1xb96-100e_ionogram.py \
                                                --out-dir output
```

Part of the output is as follows:

```shell
The information obtained is as follows:
+------------------------------+
| Information of dataset class |
+---------------+--------------+
| Class name    | Bbox num     |
+---------------+--------------+
| E             | 98           |
| Es-l          | 27           |
| Es-c          | 46           |
| F1            | 100          |
| F2            | 194          |
| Spread-F      | 6            |
+---------------+--------------+
```

> This indicate that the distribution of catogories in the dataset is unbalanced.

Reviewer: ```suggestion
This indicates that the distribution of categories in the dataset is unbalanced.
```

```shell
python tools/analysis_tools/browse_dataset.py projects/misc/ionogram_detection/yolov5/yolov5_s-v61_fast_1xb96-100e_ionogram.py \
                                              -m pipeline \
                                              --out-dir output
```

<div align=center>
<img src="https://user-images.githubusercontent.com/67947949/223914228-abcd017d-a068-4dcd-9d91-e6b546540060.png"/>

Visualization for intermediate images in the data pipeline

</div>

3. Optimize anchor size

Use the script `tools/analysis_tools/optimize_anchors.py` to obtain prior anchor box sizes suitable for the dataset.

```shell
python tools/analysis_tools/optimize_anchors.py projects/misc/ionogram_detection/yolov5/yolov5_s-v61_fast_1xb96-100e_ionogram.py \
                                                --algorithm v5-k-means \
                                                --input-shape 640 640 \
                                                --prior-match-thr 4.0 \
                                                --out-dir work_dirs/dataset_analysis_5_s
```

4. Model complexity analysis

> With the config file, the params and flops can be calculated by the script `tools/analysis_tools/get_flops.py`. Take yolov5-s as an example:

Reviewer: ```suggestion
With the config file, the parameters and FLOPs can be calculated by the script `tools/analysis_tools/get_flops.py`. Take yolov5-s as an example:
```

```shell
python tools/analysis_tools/get_flops.py projects/misc/ionogram_detection/yolov5/yolov5_s-v61_fast_1xb96-100e_ionogram.py
```

> The following output indicates that the model has 7.947G flops with the input shape (640, 640), and a total of 7.036M learnable parameters.

Reviewer: ```suggestion
The following output indicates that the model has 7.947G FLOPs with the input shape (640, 640), and a total of 7.036M learnable parameters.
```

    num_workers=train_num_workers,
    dataset=dict(
        _delete_=True,
        type='RepeatDataset',
        times=1,
        dataset=dict(
            type=_base_.dataset_type,
            indices=200,  # set indices=200，represent every epoch only iterator 200 samples
            data_root=data_root,
            metainfo=metainfo,
            ann_file=train_ann_file,
            data_prefix=dict(img=train_data_prefix),
            filter_cfg=dict(filter_empty_gt=False, min_size=32),
            pipeline=_base_.train_pipeline)))
```

**Start training**：

```shell
python tools/train.py projects/misc/ionogram_detection/yolov5/yolov5_s-v61_fast_1xb96-100e_ionogram.py
```

2. Test

Specify the path of the config file and the model to start the test:

```shell
python tools/test.py projects/misc/ionogram_detection/yolov5/yolov5_s-v61_fast_1xb96-100e_ionogram.py \
                     work_dirs/yolov5_s-v61_fast_1xb96-100e_ionogram/xxx
```

## Experiments and results

### Choose a suitable batch size

- Often, the batch size governs the training speed, and the ideal batch size will be the largest batch size supported by the available hardware.

> - When the accelerators aren't yet saturated, if the batch size doubles, the training throughput should also double (or at least nearly double). Equivalently, the time per step should be constant (or at least nearly constant) as the batch size increases.

Reviewer: ```suggestion
- If the video memory is not yet fully utilized, doubling the batch size should result in a corresponding doubling (or close to doubling) of the training throughput. This is equivalent to maintaining a constant (or nearly constant) time per step as the batch size increases.
```

Results：

| Model    | Epoch(best) | AMP   | Batchsize | Num workers | Memory Allocated | Training Time | Val mAP |
| -------- | ----------- | ----- | --------- | ----------- | ---------------- | ------------- | ------- |
| YOLOv5-s | 100(82)     | False | 32        | 6           | 35.07%           | 54 min        | 0.575   |
| YOLOv5-s | 100(96)     | True  | 32        | 6           | 24.93%           | 49 min        | 0.578   |
| YOLOv5-s | 100(100)    | False | 96        | 6           | 96.64%           | 48 min        | 0.571   |
| YOLOv5-s | 100(100)    | True  | 96        | 6           | 54.66%           | **37** min    | 0.575   |
| YOLOv5-s | 100(90)     | True  | 144       | 6           | 77.06%           | 39 min        | 0.573   |
| YOLOv5-s | 200(148)    | True  | 96        | 6           | 54.66%           | 72 min        | 0.575   |
| YOLOv5-s | 200(188)    | True  | 96        | **8**       | 54.66%           | 67 min        | 0.576   |

<div align=center>
<img width="60%" src="https://user-images.githubusercontent.com/67947949/223635966-948c8424-1ba8-4df0-92d7-079029dc1231.png">

The proportion of data loading time to the total time of each step.

</div>

> Based on the results above, we can condlude that

Reviewer: ```suggestion
Based on the results above, we can conclude that
```

> - AMP has little impact on the accuracy of the model, but can significantly reduce the memory usage while training.

Reviewer: ```suggestion
- AMP has little impact on the accuracy of the model, but can significantly reduce memory usage while training.
```

| Val mAP    | 0.507                                                                                            | 0.550                                                                                              | 0.572                                                                                                     | 0.567                                                                                                              | 0.575                                                                                       |

The results indicate that mosaic augmentation and random affine transformation can significantly improve the performance on the validation set.

#### Using pre-trained models

> To not use pre-trained weights, simply set `load_from = None` in the config file. For experiments that do not use pre-trained weights, increase the base learning rate by four times and increase the number of training epochs to 200 to ensure the model is sufficiently trained.

Reviewer: ```suggestion
If you prefer not to use pre-trained weights, you can simply set `load_from = None` in the config file. For experiments that do not use pre-trained weights, it is recommended to increase the base learning rate by a factor of four and extend the number of training epochs to 200 to ensure adequate model training.
```
